arangod/V8Server/v8-user-functions.cpp

```diff
@@ -0,0 +1,153 @@
+////////////////////////////////////////////////////////////////////////////////
+/// DISCLAIMER
+///
+/// Copyright 2014-2016 ArangoDB GmbH, Cologne, Germany
+/// Copyright 2004-2014 triAGENS GmbH, Cologne, Germany
+///
+/// Licensed under the Apache License, Version 2.0 (the "License");
+/// you may not use this file except in compliance with the License.
+/// You may obtain a copy of the License at
+///
+///     http://www.apache.org/licenses/LICENSE-2.0
+///
+/// Unless required by applicable law or agreed to in writing, software
+/// distributed under the License is distributed on an "AS IS" BASIS,
+/// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+/// See the License for the specific language governing permissions and
+/// limitations under the License.
+///
+/// Copyright holder is ArangoDB GmbH, Cologne, Germany
+///
+/// @author Wilfried Goesgens
+////////////////////////////////////////////////////////////////////////////////
+
+#include "v8-user-functions.h"
+
+//#include "Basics/Result.h"
+
+#include "v8-vocbaseprivate.h"
+#include <v8.h>
+#include "V8/v8-utils.h"
+#include "V8/v8-vpack.h"
+#include "V8/v8-conv.h"
+
+#include "VocBase/Methods/AqlUserFunctions.h"
+
+using namespace arangodb;
+
+static void JS_UnregisterAQLUserFunction(v8::FunctionCallbackInfo<v8::Value> const& args) {
+  TRI_V8_TRY_CATCH_BEGIN(isolate);
+  v8::HandleScope scope(isolate);
+
+  if (args.Length() != 1) {
+    TRI_V8_THROW_EXCEPTION_USAGE("UNREGISTER_AQL_USER_FUNCTION(<name>)");
+  }
+
+  TRI_vocbase_t* vocbase = GetContextVocBase(isolate);
+
+  std::string functionName = TRI_ObjectToString(args[0]);
+
+  Result rv = unregisterUserFunction(vocbase, functionName);
+
+  if (rv.fail()) {
+    TRI_V8_THROW_EXCEPTION(rv);
+  }
+  TRI_V8_TRY_CATCH_END;
+}
+
+static void JS_UnregisterAQLUserFunctionsGroup(v8::FunctionCallbackInfo<v8::Value> const& args) {
+  TRI_V8_TRY_CATCH_BEGIN(isolate);
+  v8::HandleScope scope(isolate);
+
+  if (args.Length() != 1) {
+    TRI_V8_THROW_EXCEPTION_USAGE("UNREGISTER_AQL_USER_FUNCTION_GROUP(<group string>)");
+  }
+
+  int deleteCount;
+  TRI_vocbase_t* vocbase = GetContextVocBase(isolate);
+
+  std::string functionFilterPrefix = TRI_ObjectToString(args[0]);
+
+  Result rv = unregisterUserFunctionsGroup(vocbase, functionFilterPrefix, deleteCount);
+
+  if (rv.fail()) {
+    TRI_V8_THROW_EXCEPTION(rv);
+  }
+  TRI_V8_RETURN(v8::Number::New(isolate, deleteCount));
+  TRI_V8_TRY_CATCH_END;
+}
+
+static void JS_RegisterAqlUserFunction(v8::FunctionCallbackInfo<v8::Value> const& args) {
+  TRI_V8_TRY_CATCH_BEGIN(isolate);
+  v8::HandleScope scope(isolate);
+
+  if (args.Length() != 1) {
```

Comment:
how can one call it with two or three parameters if it rejects all calls that do not have exactly one parameter?
the usage exception in the line below claims it can be called with one, two or three parameters.

Reply:
the .js wrapper converts the parameters to the request object compatible to the REST-Handler

Reply:
well, the args.Length() != 1 check here contradicts the usage error message in the following line.
the usage error message suggests that this function call be called with one, two or three arguments, but the args.Length() != 1 check proves this is not true.

Reply:
don't we usually generate an error message that would suit what the wrapper should throw? i.e. we use the name that the wrapper emits, not SYS_UGLY_NATIVE_NAME in all places.

Reply:
The only problem here is that your exception message says that the function can be called with one, two or three parameters, but that you will only accept a single input parameter.
If the function is supposed to be called with just a single input parameter, that's fine.
But then the usage error message should not indicate that it can be called with one to three parameters but just one.